hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestRatisUnderReplicationHandler.java

```diff
@@ -247,6 +252,131 @@ public void testInsufficientTargetsFoundBecauseOfPlacementPolicy() {
     Assert.assertEquals(1, metrics.getPartialReplicationTotal());
   }
 
+  @Test
+  public void testNoTargetsFoundBecauseOfPlacementPolicyRemoveNone() {
+    policy = ReplicationTestUtil.getNoNodesTestPlacementPolicy(nodeManager,
+        conf);
+    RatisUnderReplicationHandler handler =
+        new RatisUnderReplicationHandler(policy, conf, replicationManager);
+
+    Set<ContainerReplica> replicas
+        = createReplicas(container.containerID(), State.CLOSED, 0);
+
+    ContainerReplica shouldDelete = createContainerReplica(
+        container.containerID(), 0, IN_SERVICE, State.UNHEALTHY);
+    replicas.add(shouldDelete);
+
+    Assert.assertThrows(IOException.class,
+        () -> handler.processAndSendCommands(replicas,
+            Collections.emptyList(), getUnderReplicatedHealthResult(), 2));
+    // No commands send, as there are only 2 replicas available.
+    Assert.assertEquals(0, commandsSent.size());
+  }
+
+  @Test
+  public void testNoTargetsFoundBecauseOfPlacementPolicyNoneHealthy() {
+    policy = ReplicationTestUtil.getNoNodesTestPlacementPolicy(nodeManager,
+        conf);
+    RatisUnderReplicationHandler handler =
+        new RatisUnderReplicationHandler(policy, conf, replicationManager);
+
+    // All replicas UNHEALTHY so we do nothing.
+    Set<ContainerReplica> replicas
+        = createReplicas(container.containerID(), State.UNHEALTHY, 0, 0);
+
+    Assert.assertThrows(IOException.class,
+        () -> handler.processAndSendCommands(replicas,
+            Collections.emptyList(), getUnderReplicatedHealthResult(), 2));
+    // No commands send, as no CLOSED replicas available.
+    Assert.assertEquals(0, commandsSent.size());
+  }
+
+  @Test
+  public void testNoTargetsFoundBecauseOfPlacementPolicyRemoveUnhealthy() {
+    policy = ReplicationTestUtil.getNoNodesTestPlacementPolicy(nodeManager,
+        conf);
+    RatisUnderReplicationHandler handler =
+        new RatisUnderReplicationHandler(policy, conf, replicationManager);
+
+    Set<ContainerReplica> replicas
+        = createReplicas(container.containerID(), State.CLOSED, 0, 0);
+
+    ContainerReplica shouldDelete = createContainerReplica(
+        container.containerID(), 0, IN_SERVICE, State.UNHEALTHY);
+    replicas.add(shouldDelete);
+
+    Assert.assertThrows(IOException.class,
+        () -> handler.processAndSendCommands(replicas,
+            Collections.emptyList(), getUnderReplicatedHealthResult(), 2));
+    Assert.assertEquals(1, commandsSent.size());
+    Pair<DatanodeDetails, SCMCommand<?>> cmd = commandsSent.iterator().next();
+    Assert.assertEquals(shouldDelete.getDatanodeDetails(), cmd.getKey());
+    Assert.assertEquals(StorageContainerDatanodeProtocolProtos.SCMCommandProto
+        .Type.deleteContainerCommand, cmd.getValue().getType());
```

Comment:
Once the delete is sent, can we do another run of the under replication handler to check that the correct replica is copied after space is available?

Reply:
That should be tested elsewhere. In these unit tests, we are mocking the delete - it doesn't actually delete anything. The delete command is created and added to the commands set, rather than actually deleting anything. There should be other existing tests that check the replication of the correct sources.